smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/GradleGenerator.kt

```diff
@@ -9,78 +9,202 @@ import software.amazon.smithy.kotlin.codegen.KotlinSettings
+
 /**
- * Create the gradle build file for the generated code
+ * Generate Gradle build files
+ * @param settings [KotlinSettings] describe how the build files should be generated
+ * @param manifest Where the generated files are added
+ * @param dependencies list of [KotlinDependency]s for the project
  */
 fun writeGradleBuild(
     settings: KotlinSettings,
     manifest: FileManifest,
     dependencies: List<KotlinDependency>
 ) {
-    val writer = CodeWriter().apply {
-        trimBlankLines()
-        trimTrailingSpaces()
-        setIndentText("    ")
-        expressionStart = '#'
-    }
+    val writer = createCodeWriter()
 
-    writer.withBlock("plugins {", "}\n") {
-        if (settings.build.generateFullProject) {
-            write("kotlin(\"jvm\") version #S", KOTLIN_COMPILER_VERSION)
-        } else {
-            write("kotlin(\"jvm\")")
-        }
-    }
+    val isKmp = settings.build.generateMultiplatformProject
+    val isRootModule = settings.build.generateFullProject
 
-    if (settings.build.generateFullProject) {
-        writer.withBlock("repositories {", "}\n") {
-            write("mavenLocal()")
-            write("mavenCentral()")
-        }
-    }
+    val annotationRenderer: InlineCodeWriter = {
+        val annotations = settings.build.optInAnnotations ?: emptyList()
+        if (annotations.isNotEmpty()) {
+            val formatted = annotations.joinToString(
+                separator = ",\n",
+                transform = {
+                    "\"$it\""
+                }
+            )
 
-    writer.withBlock("dependencies {", "}\n") {
-        write("implementation(kotlin(\"stdlib\"))")
-        // TODO - can we make kotlin dependencies not specify a version e.g. kotlin("kotlin-test")
-        // TODO - Kotlin MPP setup (pass through KotlinSettings) - maybe separate gradle writers
-        val orderedDependencies = dependencies.sortedWith(compareBy({ it.config }, { it.artifact }))
-        for (dependency in orderedDependencies) {
-            write("${dependency.config}(\"#L:#L:#L\")", dependency.group, dependency.artifact, dependency.version)
+            write(formatted)
         }
     }
 
-    val annotations = settings.build.optInAnnotations
-    if (annotations != null && annotations.isNotEmpty()) {
-        writer.openBlock("val optinAnnotations = listOf(")
-            .call {
-                val formatted = annotations.joinToString(
-                    separator = ",\n",
-                    transform = {
-                        "\"$it\""
-                    }
-                )
+    val pluginsBodyRenderer: InlineCodeWriter = {
+        val pluginName = if (isKmp) "multiplatform" else "jvm"
 
-                writer.write(formatted)
+        write(
+            "kotlin(\"$pluginName\") #W",
+            { w: CodeWriter ->
+                if (isRootModule) {
+                    w.write("""version "#L"""", KOTLIN_COMPILER_VERSION)
```

Comment:
could use `w.write("version #S", KOTLIN_COMPILER_VERSION)` to get string escaping